// graphql-compose-mongoose/node8/resolvers/connection.js
"use strict";

exports.__esModule = true;
exports.default = connection;
exports.prepareCursorQuery = prepareCursorQuery;

var _getIndexesFromModel = require("../utils/getIndexesFromModel");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function connection(model, // === MongooseModel
tc, opts) {
  try {
    require.resolve('graphql-compose-connection');
  } catch (e) {
    return undefined;
  }

  const prepareConnectionResolver = require('graphql-compose-connection').prepareConnectionResolver;

  if (!prepareConnectionResolver) {
    throw new Error('You should update `graphql-compose-connection` package till 3.2.0 version or above');
  }

  const uniqueIndexes = (0, _getIndexesFromModel.extendByReversedIndexes)((0, _getIndexesFromModel.getUniqueIndexes)(model), {
    reversedFirst: true
  });
  const sortConfigs = {};
  uniqueIndexes.forEach(indexData => {
    const keys = Object.keys(indexData);
    let name = keys.join('__').toUpperCase().replace(/[^_a-zA-Z0-9]/i, '__');

    if (indexData[keys[0]] === 1) {
      name = `${name}_ASC`;
    } else if (indexData[keys[0]] === -1) {
      name = `${name}_DESC`;
    }

    sortConfigs[name] = {
      value: indexData,
      cursorFields: keys,
      beforeCursorQuery: (rawQuery, cursorData) => {
        prepareCursorQuery(rawQuery, cursorData, keys, indexData, '$lt', '$gt');
      },
      afterCursorQuery: (rawQuery, cursorData) => {
        prepareCursorQuery(rawQuery, cursorData, keys, indexData, '$gt', '$lt');
      }
    };
  });

  const _ref = opts || {},
        {
    connectionResolverName = 'connection',
    findResolverName = 'findMany',
    countResolverName = 'count',
    edgeFields,
    edgeTypeName
  } = _ref,
        sortOptions = _objectWithoutProperties(_ref, ["connectionResolverName", "findResolverName", "countResolverName", "edgeFields", "edgeTypeName"]);

  return prepareConnectionResolver(tc, {
    connectionResolverName,
    findResolverName,
    countResolverName,
    sort: _objectSpread(_objectSpread({}, sortConfigs), sortOptions),
    edgeFields,
    edgeTypeName
  });
}

function prepareCursorQuery(rawQuery, cursorData, indexKeys, indexData, nextOper, prevOper) {
  if (indexKeys.length === 1) {
    // When single index { a: 1 }, then just add to one criteria to the query:
    // rawQuery.a = { $gt|$lt: cursorValue } - for next|prev record
    const k = indexKeys[0];
    if (!rawQuery[k]) rawQuery[k] = {};

    if (indexData[k] === 1) {
      rawQuery[k][nextOper] = cursorData[k];
    } else {
      rawQuery[k][prevOper] = cursorData[k];
    }
  } else {
    // When compound index {a: 1, b: -1, c: 1 } then we should add OR criterias to the query:
    // rawQuery.$or = [
    //   { a: cursorValueA, b: cursorValueB, c: { $gt|$lt: cursorValueC } },
    //   { a: cursorValueA, b: { $gt|$lt: cursorValueB } },
    //   { a: { $gt|$lt: cursorValueA } },
    // ]
    const orCriteries = [];

    for (let i = indexKeys.length - 1; i >= 0; i--) {
      const criteria = {};
      indexKeys.forEach((k, ii) => {
        if (ii < i) {
          criteria[k] = cursorData[k];
        } else if (ii === i) {
          if (indexData[k] === 1) {
            // $FlowFixMe
            criteria[k] = {
              [nextOper]: cursorData[k]
            };
          } else {
            // $FlowFixMe
            criteria[k] = {
              [prevOper]: cursorData[k]
            };
          }
        }
      });
      orCriteries.push(criteria);
    }

    rawQuery.$or = orCriteries;
  }
}
